OpenRA.Mods.Common/Traits/BotModules/Squads/States/ProtectionStates.cs

```diff
@@ -57,11 +59,76 @@ public void Tick(Squad owner)
 			else
 			{
 				foreach (var a in owner.Units)
-					owner.Bot.QueueOrder(new Order("AttackMove", a, Target.FromCell(owner.World, owner.TargetActor.Location), false));
+				{
+					if (a.Info.HasTraitInfo<AircraftInfo>() && a.TraitsImplementing<AmmoPool>().Any())
+					{
+						if (BusyAttack(a))
+							continue;
+
+						if (!ReloadsAutomatically(a))
+						{
+							if (IsRearm(a))
+								continue;
+
+							if (!HasAmmo(a))
+							{
+								owner.Bot.QueueOrder(new Order("ReturnToBase", a, false));
+								continue;
+							}
+						}
+
+						owner.Bot.QueueOrder(new Order("AttackMove", a, Target.FromCell(owner.World, owner.TargetActor.Location), false));
+					}
+					else
+						owner.Bot.QueueOrder(new Order("AttackMove", a, Target.FromCell(owner.World, owner.TargetActor.Location), false));
+				}
 			}
 		}
 
 		public void Deactivate(Squad owner) { }
+
+		protected static bool FullAmmo(Actor a)
+		{
+			var ammoPools = a.TraitsImplementing<AmmoPool>();
+			return ammoPools.All(x => x.HasFullAmmo);
+		}
+
+		protected static bool HasAmmo(Actor a)
+		{
+			var ammoPools = a.TraitsImplementing<AmmoPool>();
+			return ammoPools.All(x => x.HasAmmo);
+		}
+
+		protected static bool ReloadsAutomatically(Actor a)
+		{
+			var ammoPools = a.TraitsImplementing<AmmoPool>();
```

Comment:
You can save a trait lookup by ordering this down.

Reply:
Those lines are directly from `AirState` in fact. Ummm... I guess you mean remove var `ammoPools`?

I can do that but I don't know why the same writing style in `AirState`, so... I don't know

Reply:
No I mean
```suggestion
			var rearmable = a.TraitOrDefault<Rearmable>();
			if (rearmable == null)
				return true;

			var ammoPools = a.TraitsImplementing<AmmoPool>();
			return ammoPools.All(ap => !rearmable.Info.AmmoPools.Contains(ap.Info.Name));
```

Reply:
Oh I see. Should I change the same one in `AirState`?